CMakeLists.txt

```diff
@@ -45,22 +46,29 @@ function(RELATIVE_PROTOBUF_GENERATE_CPP SRCS HDRS ROOT_DIR)
 
   set(${SRCS})
   set(${HDRS})
-  foreach(FIL ${ARGN})
-    message(${FIL})
-    set(ABS_FIL ${ROOT_DIR}/${FIL})
-    get_filename_component(FIL_WE ${FIL} NAME_WE)
-    get_filename_component(FIL_DIR ${ABS_FIL} DIRECTORY)
-    file(RELATIVE_PATH REL_DIR ${ROOT_DIR} ${FIL_DIR})
 
-    list(APPEND ${SRCS} "${CMAKE_CURRENT_BINARY_DIR}/${REL_DIR}/${FIL_WE}.pb.cc")
-    list(APPEND ${HDRS} "${CMAKE_CURRENT_BINARY_DIR}/${REL_DIR}/${FIL_WE}.pb.h")
+  set(GEN_PROTO_PY ${ROOT_DIR}/onnx/gen_proto.py)
+  foreach(INFILE ${ARGN})
+    message(${INFILE})
+    set(ABS_FILE ${ROOT_DIR}/${INFILE})
+    get_filename_component(FILE_DIR ${ABS_FILE} DIRECTORY)
+    get_filename_component(FILE_WE ${INFILE} NAME_WE)
+    file(RELATIVE_PATH REL_DIR ${ROOT_DIR} ${FILE_DIR})
+    set(OUTPUT_PROTO_DIR "${CMAKE_CURRENT_BINARY_DIR}/${REL_DIR}")
+
+    list(APPEND ${SRCS} "${OUTPUT_PROTO_DIR}/${FILE_WE}.pb.cc")
+    list(APPEND ${HDRS} "${OUTPUT_PROTO_DIR}/${FILE_WE}.pb.h")
 
     add_custom_command(
```

Comment:
Dependencies are not properly tracked here. In particular, changes in `*.proto.in` files would not cause a rebuild. Also, IIUC, you are calling `gen_proto.py` for every `.proto.in` file, but `gen_proto.py` generates the same set of output files. Multiple execs of `gen_proto.py` writing to the same output files would break parallel builds.

It is better to split this `add_custom_command` into two parts: running `gen_proto.py` to generate `*.proto` from `*.proto.in`, and running `protoc` on produced `*.proto` files to get `*.pb.cc` and `*.pb.h`.  `add_custom_command` invocations need both `OUTPUT` and `DEPENDS` arguments to properly track the deps.

Reply:
Yes, `gen_proto.py` actually take an argument for specific `in.proto`. I'll add that.